kroxylicious-runtime/src/main/java/io/kroxylicious/proxy/internal/clusternetworkaddressconfigprovider/BrokerAddressPatternUtils.java

```diff
@@ -110,8 +116,14 @@ static String replaceLiteralNodeId(String brokerAddressPattern, int nodeId) {
         return LITERAL_NODE_ID_PATTERN.matcher(brokerAddressPattern).replaceAll(Integer.toString(nodeId));
     }
 
-    static Pattern createNodeIdCapturingRegex(String brokerAddressPattern) {
-        var partsQuoted = Arrays.stream(brokerAddressPattern.split(Pattern.quote(LITERAL_NODE_ID))).map(Pattern::quote);
-        return Pattern.compile(partsQuoted.collect(Collectors.joining("(\\d+)")), Pattern.CASE_INSENSITIVE);
+    static String createNodeIdCapturingRegexPattern(String brokerAddressPattern, String virtualClusterName) {
+        String clusterReplaced = replaceVirtualClusterPattern(brokerAddressPattern, virtualClusterName);
+        // the -1 split limit produces an empty final item in the array if the node id replacement is at the end of the string
+        var partsQuoted = Arrays.stream(clusterReplaced.split(Pattern.quote(LITERAL_NODE_ID), -1)).map(Pattern::quote);
+        return partsQuoted.collect(Collectors.joining("(\\d+)"));
+    }
+
+    private static String replaceVirtualClusterPattern(String brokerAddressPattern, String virtualClusterName) {
+        return brokerAddressPattern.replaceAll(Pattern.quote(LITERAL_VIRTUAL_CLUSTER_NAME), virtualClusterName);
```

Comment:
Did we constrain the names of virtual clusters to be ~DNS labels?

Reply:
No, I don't think we ever did.  I think it is best that the proxy imposes this restriction unconditionally and enforces that all virtual cluster are DNS labels (in other words, fail early) even if SNI or the `$(virtualClusterName)` incantation aren't used.  This is better than a) failing late (at the point we are about to generate an invalid addresses), or worse still, let the invalid address be sent to a Kafka client (and watch them go bang 💥 in weird and confusing ways, no doubt).

Reply:
What about the case where we inject into a pattern like `$(virtualClusterName)-more-stuff`? Even if `virtualClusterName` is a valid DNS label, we may inject it into something that makes it invalid. Maybe we should require it to be injected as a label (at the start, end, or between two dots).

Reply:
it's now required by the proxy that the VCluster names are DNS labels https://github.com/kroxylicious/kroxylicious/pull/2198